File: backend/src/test/java/ca/metricalsky/winston/entity/ChannelEntityTest.java
package ca.metricalsky.winston.entity;

import ca.metricalsky.winston.database.entity.channel.ChannelEntity;
import ca.metricalsky.winston.test.TestUtils;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManager;

import java.time.OffsetDateTime;
import java.util.Set;

import static org.assertj.core.api.Assertions.assertThat;

@DataJpaTest
class ChannelEntityTest {

    @Autowired
    private TestEntityManager entityManager;

    @Test
    void persistsWithOnlyRequiredFields() {
        var channelEntity = new ChannelEntity();
        channelEntity.setId(TestUtils.randomId());

        var persistedEntity = entityManager.persistFlushFind(channelEntity);

        assertThat(persistedEntity)
                .hasFieldOrPropertyWithValue("id", channelEntity.getId())
                .hasAllNullFieldsOrPropertiesExcept("id", "keywords", "lastFetchedAt", "topics");
        assertThat(persistedEntity.getTopics())
                .isEmpty();
        assertThat(persistedEntity.getKeywords())
                .isEmpty();
    }

    @Test
    void persistsWithAllOptionalFields() {
        var channelEntity = new ChannelEntity();
        channelEntity.setId(TestUtils.randomId());
        channelEntity.setTitle(TestUtils.randomString());
        channelEntity.setDescription(TestUtils.randomString());
        channelEntity.setCustomUrl(TestUtils.randomString());
        channelEntity.setThumbnailUrl(TestUtils.randomString());
        channelEntity.setUploadsPlaylistId(TestUtils.randomId());
        channelEntity.setVideoCount(TestUtils.randomLong());
        channelEntity.setViewCount(TestUtils.randomLong());
        channelEntity.setSubscriberCount(TestUtils.randomLong());
        channelEntity.setPublishedAt(OffsetDateTime.now());
        channelEntity.setTopics(Set.of(TestUtils.randomString()));
        channelEntity.setKeywords(Set.of(TestUtils.randomString()));

        var persistedEntity = entityManager.persistFlushFind(channelEntity);

        assertThat(persistedEntity)
                .hasFieldOrPropertyWithValue("id", channelEntity.getId())
                .hasFieldOrPropertyWithValue("title", channelEntity.getTitle())
                .hasFieldOrPropertyWithValue("description", channelEntity.getDescription())
                .hasFieldOrPropertyWithValue("customUrl", channelEntity.getCustomUrl())
                .hasFieldOrPropertyWithValue("thumbnailUrl", channelEntity.getThumbnailUrl())
                .hasFieldOrPropertyWithValue("uploadsPlaylistId", channelEntity.getUploadsPlaylistId())
                .hasFieldOrPropertyWithValue("videoCount", channelEntity.getVideoCount())
                .hasFieldOrPropertyWithValue("viewCount", channelEntity.getViewCount())
                .hasFieldOrPropertyWithValue("subscriberCount", channelEntity.getSubscriberCount())
                .hasFieldOrPropertyWithValue("topics", channelEntity.getTopics())
                .hasFieldOrPropertyWithValue("keywords", channelEntity.getKeywords())
                .hasNoNullFieldsOrPropertiesExcept("properties");
    }
}

File: database/src/test/kotlin/ca/metricalsky/winston/database/entity/channel/ChannelEntityTest.kt
package ca.metricalsky.winston.database.entity.channel

import ca.metricalsky.winston.database.test.annotation.DatabaseTest
import ca.metricalsky.winston.database.test.faker.DatabaseFaker
import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.collections.shouldContainExactly
import io.kotest.matchers.date.shouldHaveSameInstantAs
import io.kotest.matchers.nulls.shouldNotBeNull
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManager

@DatabaseTest
class ChannelEntityTest(
    entityManager: TestEntityManager
): StringSpec({

    val faker = DatabaseFaker()

    "entity persists with only required fields" {
        val channel = faker.channel().minimal()

        entityManager.persistFlushFind(channel) shouldNotBeNull {
            id shouldBe channel.id
            title shouldBe null
            description shouldBe null
            customUrl shouldBe null
            thumbnailUrl shouldBe null
            uploadsPlaylistId shouldBe null
            videoCount shouldBe null
            viewCount shouldBe null
            subscriberCount shouldBe null
            publishedAt shouldBe null
            lastFetchedAt shouldNotBe null
            properties shouldBe null
            topics shouldBe emptySet()
            keywords shouldBe emptySet()
        }
    }

    "entity persists with all optional fields" {
        val channel = faker.channel().complete()

        entityManager.persistFlushFind(channel) shouldNotBeNull {
            id shouldBe channel.id
            title shouldBe channel.title
            description shouldBe channel.description
            customUrl shouldBe channel.customUrl
            thumbnailUrl shouldBe channel.thumbnailUrl
            uploadsPlaylistId shouldBe channel.uploadsPlaylistId
            videoCount shouldBe channel.videoCount
            viewCount shouldBe channel.viewCount
            subscriberCount shouldBe channel.subscriberCount
            publishedAt shouldBe channel.publishedAt
            lastFetchedAt shouldHaveSameInstantAs channel.lastFetchedAt
            properties shouldBe null
            topics shouldContainExactly channel.topics
            keywords shouldContainExactly channel.keywords
        }
    }
})
